### DISCUSSION TITLE
snakebids/resources/bids_tags.json update

### DISCUSSION THREAD

**Discussion by LouFel** [khanlab/snakebids#134] [Q&A] [answered]
Hello,

First of all, I wanted to thank you for developing your snakebids tool. I am a post-graduate intern trying to build a PET/MR pipeline and it has been really helpful. 

After running: 
```
config.update(
    snakebids.generate_inputs(
        bids_dir=config["bids_dir"],
        pybids_inputs=config["pybids_inputs"],
        derivatives=config["derivatives"],
        participant_label=config["participant_label"],
        exclude_participant_label=config["exclude_participant_label"]
    )
)
```
In my dataset, the PET images have a "trc" label (to specify the tracer used)
I noticed that the "trc" wildcard was missing in the updated config, in `input_wildcards`.

As I understand the way the inputs generation works is:

-  the `generate_inputs()` function generate the snakemake inputs,  
- `_get_lists_from_bids()` called inside populate input_path, input_lists, input_zip_lists, and input_wildcards,
- then along the process `_process_layout_wildcard()` is called which calls `read_bids_tags()`

The `read_bids_tags()` function checks compatibility for pybids as BIDS entities label differs from their key tags (e.g. Tracer -> trc), and I noticed that it is always using the `snakebids/resources/bids_tags.json` within `_process_layout_wildcard()` (function call is empty, and bids_tags.json is the default file). 

Finally, inside the `bids_tags.json` file the pair (label/key) `"tracer": "trc"` is missing, so I figured my issue might come from here. I might be wrong but adding the missing pair "tracer":"trc" inside the `bids_tags.json` should fix the problem. 

Thanks in advance for your help :smile:

**Comment by pvandyken** [comment, marked as answer]
This is actually a [`pybids`](https://github.com/bids-standard/pybids) issue, the library that does the path parsing. It will only read entities that it knows about, as defined in a config file ([see here for the default config](https://github.com/bids-standard/pybids/blob/master/bids/layout/config/bids.json)). Pybids allows custom config files, but as it turns out, `Snakebids` currently doesn't support these.

Adding basic support would be very straightforward: we can just add another parameter to `generate_inputs` called `config` and pass it transparently on to `BIDSLayout`. I'm going to make an issue from this. @tkkuehn , @akhanf, @kaitj, any thoughts?

**Reply by akhanf** [reply]
I agree would be great to have support for custom configs to pass along to pybids, and that seems like an easy addition to make.

Since we would be having to update the pybids config to add the missing entities in any case, may as well also open a PR (or see if there is one in progress) to update the entities in pybids too, since I believe PET has been in the bids standard for some time now already..

**Reply by akhanf** [reply]
another related question is whether we want to refactor a bit to use the pybids config json format instead of our own bids _tags.json (or have some way of one generating the other).. as otherwise we have two different configs to maintain the same thing.

**Reply by pvandyken** [reply]
My understanding of `bids_tags.json` is that it's basically just a thesaurus. The relevant line is:
```py
# input_generation.py:344
tag = bids_tags[wildcard_name] if wildcard_name in bids_tags else wildcard_name
```
So if a wildcard_name is in `bids_tags`, it will get changed from the `Pybids` version to the actual name found in the filename (i.e. `acquisition -> acq`). Otherwise it will get passed through transparently.

Unfortunately, that information is somewhat buried in the [pybids config file](https://github.com/bids-standard/pybids/blob/master/bids/layout/config/bids.json). We'd have to try to parse the `"pattern"` field to get the data:
```json
{
    "name": "acquisition",
    "pattern": "[_/\\\\]+acq-([a-zA-Z0-9]+)"
},
```
Probably possible, but not super clean.

The upshot is we only need entries in `bids_inputs.json` for entries that have alternate names.

**Reply by kaitj** [reply]
I like the idea of having custom configs alongside to be able to add any missing entities, though I think there should either be some check in place that it still conforms to bids (not sure off the top of my head how this would work and it can get tricky with derivatives) or a message to indicate that a custom config is used (if not already printed from `pybids`). 

As for refactoring, ~I don't see any reason why we shouldn't and~ it does sounds like a good idea if the two configs are ultimately maintaining the same thing. (Edit: Might need to think about this one a bit more given Peter's reply above)

**Reply by LouFel** [reply]
Thank you all for your quick responses, I will try to open a PR tomorrow to implement what @pvandyken described as basic support for the config file.

**Reply by pvandyken** [reply]
@LouFel Already on it! Should have it by this afternoon.

**Reply by LouFel** [reply]
Never mind then, thank you @pvandyken !

**Reply by pvandyken** [reply]
@LouFel The update's been merged. You can use it by installing the latest version of `snakebids` directly off snakebids. You'll need to provide your own custom config file for pybids; you should just be able to base it off the default ones.

**Reply by LouFel** [reply]
Thank you @pvandyken , it's perfectly working. Just a little warning if you want to use this functionality in the future:
In the pybids_config.json file, the name of the entities specified seems to matter (case by case), let me explain with an example:
```
{
    "name": "tracer",
    "pattern": "[_/\\\\]+trc-([a-zA-Z0-9]+)"
}
```
```
  pet:
    filters:
      suffix: 'pet'
      extension: '.nii.gz'
      datatype: 'pet'
    wildcards:
      - subject
      - session
      - tracer
      - reconstruction
      - run
```
Here with the tracer entity added in the pybids_config.json file, with the wildcard entry specified in the config.yml file as tracer, it returns this error:
```
'NoneType' object is not subscriptable
File "PATH_TO_PROJECT/workflow/Snakefile", line 13, in <module>
File "PATH_TO_VIRTUALENV/lib/python3.9/site-packages/snakebids/core/input_generation.py", line 187, in generate_inputs
File "PATH_TO_VIRTUALENV/lib/python3.9/site-packages/snakebids/core/input_generation.py", line 478, in _get_lists_from_bids
File "PATH_TO_VIRTUALENV/lib/python3.9/site-packages/snakebids/core/input_generation.py", line 385, in _process_layout_wildcard
```
Whereas with an entity like reconstruction that follows exactly the same logic:
```      
{
    "name": "reconstruction",
    "pattern": "[_/\\\\]+rec-([a-zA-Z0-9]+)"
}
```
With the wildcard defined as reconstruction in the config.yaml file, it works perfectly fine...
Simply replacing the name and wildcard by "trc" fix the issue for the tracer entity. So it seems to me that the problem is entity related, but I don't really understand why.
Maybe I am missing something, but I let you know just in case.

**Reply by akhanf** [reply]
Ah yes that's what we use bids_tags.json for.. I've made a quick PR to update that too #139

**Reply by pvandyken** [reply]
For future reference, the issue here is that, in your tracer example, the name of the entity: `tracer`, and the actual tag found on the file path: `trc`, are different. Snakebids has no way of knowing the two are meant to be equivalent, other than the `bids_tags.json` file which catalogues those equivalencies. Unfortunately, there's no easy way for users to update the `bids_tags.json` file directly in their installed library, which obviously isn't reproducible. We've discussed fixing this before though... @akhanf, I have some ideas, I'm going to open a new discussion item for them.

**Comment by akhanf** [comment]
Ah yes you're right @pvandyken, my memory of bids_tags.json wasn't very good! 
I take back my suggestion to harmonize that element across snakebids and pybids..
